[Jonathan Lee] iP

yeppog opened the pull request:
# Duke

> Beep boop

Duke chat bot written in **java**.

It can do the following:
- Add Tasks
- View Tasks
- Remove Tasks
- Set Deadlines
- Add Events

The `Duke` class can be __instantiated__ with an *optional* `config` as well as a `DukeDB`.
For example, the main method holds:
```java
class Main {
    public static void main(String[] args){
        DukeDb database = new DukeDB("path-to-database-txt-file");
        Duke bot = new Duke(database);
    }
}
```

## Plans ✨
- [ ] **Add** JavaFX
- [ ] Add more assertions


## Releases
View [here](https://github.com/yeppog/ip/releases).

## Setup
1. Clone this repository using `git clone https://github.com/yeppog/ip.git`
2. Build this project using IntelliJ and run.

Commit:
Add Gradle support

yeppog committed:
Add echo functionality

Seperated the Duke instance to a Main class

yeppog committed:
Add list storage functionality

Make Duke to be instance based and removed all static methods

yeppog committed:
Add Task class with mark complete feature

Features basic error handling for handling errenous user input.

yeppog committed:
Add Task class Javadocs

yeppog committed:
Add Todos, Events and Deadlines

Update printed messages for successful task add.

yeppog committed:
Remove enhanced switch statement

Java 15 was somehow the default, which allowed for enhanced switch. Switch back to SDK 11

yeppog committed:
Add test files

Run simple test on the input and output and check the diff

yeppog committed:
Add DukeException error handling class

Currently tries and catches errrenous methods

yeppog committed:
Add delete function, add enums

Partial refactor to use Java Enums while parsing the command.

yeppog committed:
Add DukeActions interface

This interface is used as a lambda in DukeCommands

yeppog committed:
Add enums to be used instead of switch statements

Enums contain their respective methods

yeppog committed:
Add Parser class to cleanly sort the input

HashMap implementation to store keywords to values

yeppog committed:
Remove redundant commands

Deprecated commands removed as enums replace the commands

yeppog committed:
Remove throw exception on listen() command

Accidentally placed in the code

yeppog committed:
Docs: Add java docs

Cleaned up unwanted code

yeppog committed:
chore: cleanup

Removed deprecated javadocs

yeppog committed:
Add javadocs

Cleanup existing code

yeppog committed:
Add enums to Parser

Prevents accidental parsing of non registered commands. Currently takes them as text input only.

yeppog committed:
Update test files

Update new input and expected output.

yeppog committed:
Fix incorrect message

Deadline should print missing /by and not /at

yeppog committed:
Add simple DB handler class

Add DB handler to read, create and delete files

yeppog committed:
Update DukeDB

Add commands to output the desired store string. \n Add reader methods to read the db

yeppog committed:
Add date handler

Add partial configurator

yeppog committed:
Merge branch branch-Level-8 to master

yeppog committed:
Fix database parsing deadlines

yeppog committed:
Add more OOP

Add classes UI and DukeTaskList.
Increase interaction between objects with more method calls.
Increase abstraction for methods.

yeppog committed:
Add packages

- Add Duke package for main file
- Add Duke.util package for util files

yeppog committed:
Add more packages

- Abstract enum commands into seperate action class files
- Move DukeDB and DukeException to util package
- Add task package

yeppog committed:
Add basic Junit test

- Test if Duke can be initialised
- Test if Event returns the correct /at location

yeppog committed:
Fix FileNotFoundException error

Move blocks around to accomodate all cases

yeppog committed:
Update gitignore

Chore
- Add /data and .jar output to gitignore

yeppog committed:
Add find command

- Fix DeleteCommand not printing the result.
- Find command matches strings in the description.

yeppog committed:
Add JavaDocs

Add Javadocs for all java classes

yeppog committed:
Fix boolean method

Fix boolean method to comply with coding convention for DukeActions

yeppog committed:
Merge branch 'branch-A-CodingStandard' into master

AllardQuek commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,73 @@
+package duke;
+import duke.command.DukeCommands;
+import duke.util.*;
```
Should more specific imports be used here instead?

yeppog replied:
addressed in ca26cb67916444729632edba7164a8834ff5f6e2

AllardQuek commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,73 @@
+     * Simple string array to store inputs
+     **/
+    private final DukeTaskList list;
+
```
Not sure if the blank line here is necessary?

yeppog replied:
fixed in ca26cb67916444729632edba7164a8834ff5f6e2

AllardQuek commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,73 @@
+    public Duke(DukeDB database, DukeConfig config) {
+        this.ui = new Ui();
+        this.database = database;
+        this.list = new DukeTaskList(database.load().orElse(new ArrayList<>()));
```
Looks like a good use of Optional here!

AllardQuek commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,73 @@
+
+
+    /**
+     * Simple formatting tool to be used when printing commands
```
Could consider phrasing as an action instead? e.g. Prints ...

aryagiat replied:
I notice the same issue as well on other public methods.

AllardQuek commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,73 @@
+     */
+    public void listen(){
+        Scanner scanner = new Scanner(System.in);
+        this.ui.greet(this.list.getSize());
```
Could consider omitting the `this` keyword here? My tutor mentioned to avoid unnecessary usage of `this`, but I would personally not mind using `this` as well unless it gets too cluttered.

yeppog replied:
hmmm, I would think `this` helps to keep things cleaner as it can be easily recognized as an object instance variable as compared to the other variables.

but I do agree that it might be unnecessary as it doesn't serve any purpose. 
however, I will keep `this` for now

AllardQuek commented on src/main/java/duke/command/DukeCommands.java:
```diff
@@ -0,0 +1,51 @@
+/**
+ * The DukeCommands enum encapsulates the logic for commands.
+ */
+public enum DukeCommands {
```
Should this be singular instead? When instantiating a command it sounds a bit weird to say `DukeCommands command`

AllardQuek commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,73 @@
+            Optional<DukeCommands> prefix = DukeCommands.getCommand(scannedLine.split(" ")[0]);
+            DukeCommands command = prefix.orElseGet(() -> DukeCommands.INVALID);
+            try {
+                terminate = command.action.runAndCanContinue(Parser.parseCommand(scannedLine), this.list, this.database,
```
If this line is exceeding the max length (100 characters), could consider separating the parameters on multiple lines?

AllardQuek commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,73 @@
+    public void listen(){
+        Scanner scanner = new Scanner(System.in);
+        this.ui.greet(this.list.getSize());
+        boolean terminate = true;
```
Perhaps make the boolean variable sound more like a boolean? e.g. toTerminate

AllardQuek commented on src/main/java/duke/command/ByeCommand.java:
```diff
@@ -0,0 +1,35 @@
+     * @throws DukeException When erroneous inputs are given.
+     */
+    @Override
+    public boolean runAndCanContinue(Map<String, String> map, DukeTaskList list, DukeDB database, DukeConfig config, Ui ui)
```
Not too sure what exactly this does, but could there be a better naming for this? Maybe "runCommandSuccessfully"?

yeppog replied:
i intended it to dictate the following two things:
1. to **run** the command, since its a command class
2. to return a boolean to check if the program can continue running (which does in all cases other than when the `bye` command is issued)

would modifying the javadocs help to clarify this, or what other names could I possibly use that can encapsulate the above meaning?

stephanie-csy replied:
Maybe you can split it into 2 methods? The first one could be a `void` method `runCommand` that just runs the command, and at the end you decide whether you continue running it using a `boolean` method `canContinue`.

AllardQuek commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,81 @@
+     * @param isDone Whether the task has been completed.
+     */
+    public Deadline(String description, LocalDateTime by, boolean isDone) {
+        super(description,
```
Any reason why you have the parameters on separate lines instead of a single line?

yulonglim replied:
Java coding standard for 2103 does not specify that params should be written in 1 single line. Even in examples given they are written in multiple lines.

yeppog replied:
Agree with @yulonglim, but I also do acknowledge that it looks kind of awkward given such a short line

AllardQuek commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,81 @@
+ */
+public class Deadline extends Task {
+
+    protected LocalDateTime by;
```
Could there be a better name for this? Not sure if `by` might be too vague because it can have different meanings.

AllardQuek commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,81 @@
+
+        if (by.getHour() == 0 && by.getMinute() == 0) {
+            return by.format(DateTimeFormatter.ofPattern("dd/MM/yyyy"));
+        } else {
+            return this.saveDate();
+        }
```
Maybe could consider omitting the `else` keyword? Might look a bit cleaner if you prefer

AllardQuek commented on src/main/java/duke/util/DukeConfig.java:
```diff
@@ -0,0 +1,53 @@
+     * @return DukeConfig object
+     */
+    public static DukeConfig of (String config) {
+        return DukeDateConfig.getDukeDateConfig(config).map(DukeConfig::new).orElseGet(DukeConfig::new);
```
Could consider separating on multiple lines?

AllardQuek commented on src/main/java/duke/util/DukeDB.java:
```diff
@@ -0,0 +1,168 @@
+            return Optional.of(arr);
+        } catch (FileNotFoundException e) {
+            Duke.printMsg(e.toString());
+        } catch (Error f) {
```
Any reason why the error is named f?

AllardQuek commented on src/main/java/duke/util/DukeDB.java:
```diff
@@ -0,0 +1,168 @@
+        } else {
+            String description = splitCommand[1];
+            boolean isDone = splitCommand[2].equals("1");
+            return Optional.of(new Todo(description,
+                    isDone));
```
Not sure if the indentation of the second line should be more? Maybe align with the first parameter?

AllardQuek replied:
Not sure if the indentation of the second line should be more? Maybe align with the first parameter?

AllardQuek commented on src/main/java/duke/util/DukeDB.java:
```diff
@@ -0,0 +1,168 @@
+    }
+
+    private Optional<Task> readEvent(String[] splitCommand) {
+        if (splitCommand.length != 4) {
```
Could consider creating a variable for the length to avoid magic numbers?

AllardQuek commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,157 @@
+
+        String key = splitted[0];
+        StringBuilder arguments = new StringBuilder();
+        for (int i = 1;
+             i < splitted.length;
+             i++) {
```
Is there a reason for splitting onto multiple lines?

AllardQuek commented on src/test/java/duke/task/EventTest.java:
```diff
@@ -0,0 +1,16 @@
+public class EventTest {
+    @Test
+
+    public void initTest() {
```
Not sure if the naming convention for tests is followed?

AllardQuek commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,157 @@
+        }
+
+        public static boolean isValid(String test) {
+            return Arrays.stream(ValidParams.values())
+                    .anyMatch(x -> x.value.equals(test));
```
Could consider aligning `anyMatch` with `stream`?

yeppog replied:
dont see much of an issue since there are 8 spaces for the wrapped line

```java
return Arrays
        .stream(...)
        .anyMatch()
```
looks rather awkward as well since the return line is rather short

AllardQuek commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,157 @@
+            arguments.deleteCharAt(arguments.length() - 1);
+            inputMap.put(key,
+                    arguments.toString());
+        } else {
+            if (!inputMap.containsKey(key)) {
```
Could consider making it an `else if` instead?

aryagiat commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,73 @@
+    /**
+     * Simple string array to store inputs
+     **/
+    private final DukeTaskList list;
```
Should a constant variable name be in lowercase or uppercase?

aryagiat commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,73 @@
+     **/
+    private final DukeTaskList list;
+
+    private final DukeDB database;
+
+    private final DukeConfig config;
+
+    private final Ui ui;
```
Same here, should a constant variable name be in lowercase?

aryagiat commented on src/main/java/duke/Main.java:
```diff
@@ -0,0 +1,12 @@
+
+import duke.util.DukeDB;
+
+public class Main {
+    public static void main(String[] args) {
```
Should there be comments for public classes and methods?

aryagiat commented on src/main/java/duke/command/ByeCommand.java:
```diff
@@ -0,0 +1,35 @@
+package duke.command;
+
+import duke.Duke;
+import duke.util.DukeDB;
+import duke.util.DukeException;
+import duke.util.DukeConfig;
+import duke.util.DukeTaskList;
+import duke.util.Ui;
```
I like the alphabetical arrangement of the import statement 👍

aryagiat commented on src/main/java/duke/command/ByeCommand.java:
```diff
@@ -0,0 +1,35 @@
+ * The ByeCommand class encapsulates the run method for the bye command.
+ */
+public class ByeCommand implements DukeActions {
+
+
```
Is the spacing needed?

aryagiat commented on src/main/java/duke/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,46 @@
+     * @throws DukeException When erroneous inputs are given.
+     */
+    @Override
+    public boolean runAndCanContinue(Map<String, String> map, DukeTaskList list, DukeDB database, DukeConfig config, Ui ui)
```
Does the naming sound boolean-like? maybe try: `isCanRunAndContinue`

yeppog replied:
the boolean is however not asserting if it can run. it is however checking if we should continue running the bot

aryagiat commented on src/main/java/duke/util/DukeTaskList.java:
```diff
@@ -0,0 +1,72 @@
+public class DukeTaskList {
+    private final ArrayList<Task> tasklist;
+
+    public DukeTaskList() {
```
Should there be javadocs comments for all public classes/methods?

aryagiat commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,157 @@
+     */
+    public static Map<String, String> parseCommand(String text) {
+        HashMap<String, String> inputMap = new HashMap<>();
+        String[] splitted = text.split("\\s");
```
Should the variable name of a list be in plural form?

aryagiat commented on src/main/java/duke/util/Ui.java:
```diff
@@ -0,0 +1,118 @@
+    /** Final value for the border */
+    final private static String LINES = "===============================================";
+
+    /**
+     * Basic constructor
+     */
+    public Ui (){}
```
Is this redundant as the default constructor is automatically created? Is there a space in between the closing bracket `)` and `{`?

yulonglim reviewed (commented):
<ignore>

yulonglim commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,38 @@
+     * @param text The text to be printed.
+     */
+    public static void printMsg(String text) {
+        String bar = "===============================================";
```
bar seems to be a constant, might be better to use all caps to name

yulonglim commented on src/main/java/duke/command/DukeActions.java:
```diff
@@ -1,10 +1,6 @@
-import duke.util.DukeConfig;
-import duke.util.DukeTaskList;
-import duke.util.Ui;
+import duke.util.*;
```
maybe revert this change? not encouraged to use * import

yeppog replied:
fixed in ca26cb67916444729632edba7164a8834ff5f6e2

yeppog committed:
Update styling

Change variable and method names to fit the style guide.

yeppog committed:
Fix styling issues

Update imports

yeppog committed:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2122S1/ip into add-gradle-support

yeppog committed:
Add gradle

yeppog committed:
Add checkstyle config

stephanie-csy commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,85 @@
+    }
+
+
+    /**
+     * Formats print commands with borders.
+     *
+     * @param text The text to be printed.
+     */
+    public static void printMsg(String text) {
+        String bar = "===============================================";
+        System.out.printf("%s\n%s\n%s\n",
+                bar,
+                text,
+                bar);
+    }
```
Should this be in the Ui class instead?

stephanie-csy commented on src/main/java/duke/command/DukeActions.java:
```diff
@@ -0,0 +1,32 @@
+/**
+ * The DukeActions interfaces encapsulates the run method for the any command.
+ */
+public interface DukeActions {
```
Is there a reason you use an interface instead of an abstract class? Since the example given on the iP website was to have an abstract class Command.

stephanie-csy commented on src/main/java/duke/util/DukeDB.java:
```diff
@@ -0,0 +1,171 @@
+import java.util.Optional;
+import java.util.Scanner;
+
+public class DukeDB {
```
Consider adding javadocs for this class and its methods! Because they are not very straightforward

stephanie-csy reviewed (commented):
Clean code that does not have any serious coding standard violations. Can consider adding more javadocs to make code even more readable. Naming conventions are generally well followed, except for the `runAndCanContinue` one. Interesting use of streams! Well done in general 👍

yeppog committed:
Add GUI

Add simple GUI for duke. Removed CLI

yeppog committed:
Update build.gradle

Include JavaFX dependencies

yeppog committed:
Merge branch 'branch-Level-10' into master

yeppog committed:
Update Gradle

Add Gradle settings back.

yeppog committed:
Add assertions to DukeDB

DukeDB handles raw user input string to do file operations, which can be
buggy.

Adding assertions will allow us to prevent specific run-time related
errors.

Assertions are easy to implement

Added a new constructor to prevent calling "null" on the constructor of
DukeDB to default to the standard DB path.

yeppog committed:
Merge tag 'A-Assertions' into branch-A-Assertions

yeppog committed:
Add CI

yeppog committed:
Update code quality

yeppog committed:
Add C-Help to Duke

Allow for a help command to display the available commands to the user.

yeppog committed:
Merge branch 'branch-C-Help' into master

yeppog committed:
Merge branch 'branch-A-CodeQuality' into master

yeppog committed:
Merge branch 'branch-A-Assertions' into master

yeppog committed:
Merge branch 'branch-A-CI' into master

yeppog committed:
Fix checkstyle issue

yeppog committed:
Add gettr for test

yeppog committed:
Add user docs

yeppog committed:
Fix checkstyle issue

yeppog committed:
Update CI

yeppog committed:
Update gradle.yml

yeppog committed:
Update doc image name
